### PR TITLE
1542: fix regex issues

### LINKED ISSUE #1542: (JS-S001) Regular Expression Denial of Service (ReDoS)
## Description
The Regular expression Denial of Service (ReDoS) is a Denial of Service attack that exploits the fact that most Regular Expression implementations may reach extreme situations that cause them to work very slowly (exponentially related to input size).

## Occurrences
There are 4 occurrences of this issue in the repository.

See all occurrences on DeepSource &rarr; [deepsource.io/gh/kodadot/nft-gallery/issue/JS-S001/occurrences/](https://deepsource.io/gh/kodadot/nft-gallery/issue/JS-S001/occurrences/)

image

### PR DESCRIPTION
**Pull request by rhaicode**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇  _ Do a quick check before the merge. 

### PR type
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

### Before submitting Pull Request, please make sure:
- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted screenshot of demonstrated change in this PR

### Optional
- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything works
- [ ] I found edge cases

### What's new?
- [x] PR closes #1542
- [x] Updated regex tests

### Had issue bounty label ?
- [x] Fill up your KSM address: [Payout](https://kodadot.xyz/transfer/?target=EJ4LJsZAefUW5pwokk5G5b96CpHaABiXXaEUpdBakP8oYJr)

### Community participation
- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot
- [ ] My fix has changed **something** on UI, a screenshot for others, is best to understand changes.

### CODE REVIEW

File: i18n.ts [hunk 1]
```diff
@@ -12,13 +12,13 @@ const md = MarkdownIt({
...
   const allLangDataFile = 'all_lang.json'
-  const locales = require.context('./langDir', true, /[A-Za-z0-9-_,\s]+\.json$/i)
+  const locales = require.context('./langDir', true, /[A-Za-z0-9-_,\s]{1,8}\.json$/i)
   const messages: LocaleMessages = {}
   locales.keys().forEach(key => {
     if (key === allLangDataFile) {
       return
     }
-    const matched = key.match(/([A-Za-z0-9-_]+)\./i)
+    const matched = key.match(/([A-Za-z0-9-_]{1,8})\./i)
```
Comment (vikiival) [inline i18n.ts:21]: why 8?

File: components/rmrk/utils.ts [hunk 2]
```diff
@@ -228,14 +228,12 @@ export const getView = (rmrkString: string): RmrkMint | RmrkView | null => {
 }
 
 export const unwrap = (rmrkString: string): any | null => {
-  const rr = /{.*}/
-  const match = rmrkString.match(rr)
-
-  if (!match) {
+  try {
```
Comment (vikiival) [inline components/rmrk/utils.ts:231]: It seems is not correct

```ts
a = 'RMRK::MINT::1.0.0::{a:1, b:2, c: "dr.dre" }'
const rr = /{.*}/
const x = a.match(rr)
console.log(x)
// will print: ["{a:1, b:2, c: \"dr.dre\" }"]
```

**Review by vikiival — COMMENTED** [review]
Not sure how unwrap is used but unless you do not split RMRK string by squares this impl will fail

File: components/rmrk/utils.ts (hunk 2 above)
Comment (rhaicode) [reply components/rmrk/utils.ts:231]: okay. i thought the rmrkString is already an object. in this case i would have to reevaluate again an approach for this one

File: i18n.ts (hunk 1 above)
Comment (rhaicode) [reply i18n.ts:21]: since the longest name of a json file in langDir is all_lang, which also happens to have 8 characters

Comment (rhaicode) [reply i18n.ts:21]: having + in this case would be abused in NFA (non deterministic finite automation) so i tried to only evaluate the number of necessary characters

**Comment by yangwao** [comment]
let me know when it's done! 
/giphy security dog

**Comment by yangwao** [comment]
@rhaicode let review on this if you are working on this one :)

**Comment by rhaicode** [comment]
Hi @yangwao, I am closing this PR since I have to fully verify whether the code that I'm going to supply works or not and I feel not really confident working on this one. I'll just grab other issues instead.

**Pull request closed without merging**